### PR TITLE
[Issue-390]:  git workflow to build and push docker image to docker hub

### PR DESCRIPTION
**Pull request by ThirumlaDevi**
-   Please make sure the below checklist is followed for Pull Requests.
-   [X] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

### Description 
Issue link : [ISSUE-390](https://github.com/jhipster/jhipster-dotnetcore/issues/390)
The following contains information/decision made in newly introduced github workflow to build and push docker image
- The workflow is configured to be triggered only when a PR is merged.
- The dockerhub username, token and repository must be configured as a github repository secret for the workflow to work. Github secret is recommended to be used in github workflows for the following reasons :
   - Secrets are environment variables that are encrypted and only exposed to selected actions. 
   - Secrets can only be configured by GitHub repository project owners. Hence, secrets cannot be modified by unauthorised individual.
- On docker push action, i have configured the github workflow to push image with latest tag and current version tag, this way when a new version/change is made, then that change will be your latest tag.

### Testing Done 
- For testing purpose, I modified triggering of GitHub action to be done on commit push
  - Please check this branch's [commit history](https://github.com/ThirumlaDevi/jhipster-dotnetcore/commits/ISSUE_390_git_workflow_to_build_and_push_docker_img), it'll clearly indicate the above action

- Created a sample private repository in my docker hub account for testing if docker image gets pushed to this repository
image

- Configured the secrets to my forked GitHub repository settings/secrets page
image

- Checked if GitHub action is triggered and that the triggered job is executed successfully
image

- Checked if image gets pushed with proper tags
_**Note: The digest value of image having the latest tag and the versioned tag is the same. As new changes are published the image having latest tag will be in par with the newly released tag**_
image

- PS: I deleted the images, after testing was done
image

### CODE REVIEW

File: .github/workflows/build-and-push-docker-image.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+name: build and push docker image
+
+on:
+  pull_request:
+    types: closed
```
Comment (nicolas63) [inline .github/workflows/build-and-push-docker-image.yml:5]: Perhaps we can only push the image when a new tag has been created

**Review by nicolas63 — CHANGES_REQUESTED** [review]
Can you complete the documentation here (https://github.com/jhipster/jhipster-dotnetcore#using-docker) ?

**Comment by nicolas63** [comment]
@jdubois Can you give me access to dockerhub ? Or just complete secrets DOCKERHUB_REPOSITORY, DOCKERHUB_TOKEN, DOCKERHUB_USERNAME ?

**Comment by jdubois** [comment]
The best solution is to fill up the tokens (a bit more secure), @pascalgrimaud I believe this has changed for the main images, are you doing the same?

**Comment by pascalgrimaud** [comment]
Just read quickly this ticket, I'm not sure it's the right way to do this.

Everything can be done at Docker Hub, so we'll have automated builds as other images (jhipster online, jhipster registry etc)

I can configure there.
Let me do this, if you confirm it.

**Comment by nicolas63** [comment]
@pascalgrimaud, you can trigger build directly with docker hub ? i'm ok with this

**Comment by pascalgrimaud** [comment]
@nicolas63 : exactly. What is done for jhipster-registry:
- trigger when a tag is created, the image name is: jhipster/jhipster-registry:tagname, for example jhipster/jhipster-registry:v6.4.0
- trigger when a tag is created too, the image is: jhipster/jhipster-registry:latest

We had jhipster/jhipster-registry:master too, but I deleted the trigger yesterday because I don't think people use it
And I didn't want to create jhipster/jhipster-registry:main

I can do the same for dotnet.
Is it ok for you ?

**Comment by nicolas63** [comment]
yes, it's ok for me thanks for your help ! 

@ThirumlaDevi Thanks for your work but sorry i close this PR the pascalgrimaud solution seems better.

**Comment by pascalgrimaud** [comment]
Everything should be configured at Docker Hub - See https://hub.docker.com/repository/docker/jhipster/jhipster-dotnetcore

Capture d’écran de 2020-10-25 09-30-56

I will probably remove 'main' trigger, unless the `jhipster/jhipster-dotnet:main` will be used a lot. Just tell me what you want.
You could add a build, to test the image:
- https://github.com/jhipster/jhipster-registry/blob/main/.github/workflows/docker.yml
- it's not the same, but the idea is to build the image

**Comment by pascalgrimaud** [comment]
@nicolas63 : IMO, I think this PR can be changed to test docker build, with: `docker build -t jhipster/jhipster-dotnetcore:ci ./docker/`

**Comment by pascalgrimaud** [comment]
The first image has been published:
- https://hub.docker.com/repository/docker/jhipster/jhipster-dotnetcore/tags?page=1
- try to test it: `docker pull jhipster/jhipster-dotnetcore:main`

**Comment by nicolas63** [comment]
I think you can leave the build 'main' for now. If nobody uses it we can delete it. 
@pascalgrimaud thank you for your advice 
@ThirumlaDevi Do you whant change your PR to test docker build ?

**Comment by pascalgrimaud** [comment]
@nicolas63 : don't hesitate to create a 'test' tag, to be sure that docker hub trigger a build, and create latest + test image : it can take sometimes some hours, as the Organization JHipster can only build 1 image at the same time
Then, ping me if it doesn't work

**Comment by ThirumlaDevi** [comment]
@nicolas63 I was going to suggest the same, where you can link github repository to a docker hub repo to trigger image builds in docker hub, but when i tried it there were build failures ( I assumed because i linked a forked repository). But, it's awesome that you were able to solve this in your end.
To answer [your question ](https://github.com/jhipster/jhipster-dotnetcore/pull/425#issuecomment-716112990), i believe you want me to change the workflow to check if docker image build is successful, when a push commit event is made on any branch in  jhipster-dotnetcore repo?

**Comment by nicolas63** [comment]
> To answer [your question ](https://github.com/jhipster/jhipster-dotnetcore/pull/425#issuecomment-716112990), i believe you want me to change the workflow to check if docker image build is successful, when a push commit event is made on any branch in jhipster-dotnetcore repo?

Just test if the image build no like suggest by pascalgrimaud with something like this `docker build -t jhipster/jhipster-dotnetcore:ci ./docker/` 
With this test we ensure that docker image is not broken

**Comment by ThirumlaDevi** [comment]
@nicolas63 I have made the requested changes:)
I would like to make a suggestion. Since you want to check/test if docker image is not broken, it'll require more changes than just testing if docker image build is successful. We also need to extend this workflow in future to check the following:

- Start a container using the image.
- Run end-to-end tests across this container, to check if written code is imported and run

I found the below link useful on how to test if docker image is broken?
https://pythonspeed.com/articles/test-your-docker-build/

File: .github/workflows/build-and-push-docker-image.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+name: build docker image
+
+on: push
```
Comment (nicolas63) [inline .github/workflows/build-and-push-docker-image.yml:3]: We can execute also on pull_request

**Review by nicolas63 — CHANGES_REQUESTED** [review]

**Comment by nicolas63** [comment]
I agree with you, can you open up the issue so that we don't forget ?

File: .github/workflows/build-and-push-docker-image.yml (hunk 2 above)
Comment (ThirumlaDevi) [reply .github/workflows/build-and-push-docker-image.yml]: Done

**Review by nicolas63 — APPROVED** [review]

**Comment by nicolas63** [comment]
🎉 thanks for your contribution @ThirumlaDevi

**Comment by ThirumlaDevi** [comment]
@nicolas63 I have made the requested changes and i have also raised the [issue](https://github.com/jhipster/generator-jhipster/issues/12841)

**Pull request merged**